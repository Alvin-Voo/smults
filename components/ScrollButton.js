import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button } from 'semantic-ui-react';

class ScrollButton extends Component {
  state={
    intervalId: 0,
  }

  scrollStep=() => {
    const { intervalId } = this.state;
    const { scrollStepInPx } = this.props;
    if (window.pageYOffset === 0) {
      clearInterval(intervalId);
    }
    window.scroll(0, window.pageYOffset - scrollStepInPx);
  }

  scrollToTop=() => {
    const { delayInMs } = this.props;
    const intervalId = setInterval(this.scrollStep, delayInMs);
    this.setState({ intervalId });
  }

  render() {
    return (
      <Button
        className="scrollButton"
        title="Back to top"
        icon="angle double up"
        onClick={this.scrollToTop}
      />
    );
  }
}

ScrollButton.propTypes = {
  scrollStepInPx: PropTypes.number.isRequired,
  delayInMs: PropTypes.number.isRequired,
};

export default ScrollButton;
