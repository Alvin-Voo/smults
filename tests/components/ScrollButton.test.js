import React from 'react';
import { shallow } from 'enzyme';
import ScrollButton from '../../components/ScrollButton';


let wrapper;

beforeEach(() => {
  wrapper = shallow(<ScrollButton scrollStepInPx={50} delayInMs={16.66} />);
});

test('should render correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

describe('Scroll Up click function tests', () => {
  jest.useFakeTimers();

  test('should trigger scrollToTop function correctly', () => {
    wrapper.simulate('click');
    expect(setInterval).toHaveBeenCalledTimes(1);
    expect(setInterval).toHaveBeenCalledWith(wrapper.instance().scrollStep, 16.66);
    expect(wrapper.state('intervalId')).toBeGreaterThan(0);
  });

  test('should trigger scrollStep function correctly', () => {
    // the test case before this has called click once, so clear before hand
    jest.clearAllTimers();
    jest.clearAllMocks();

    window.pageYOffset = 1000;
    window.scroll = jest.fn((initVal, offsetVal) => {
      window.pageYOffset = offsetVal;
    });
    wrapper.simulate('click');

    const intervalId = wrapper.state('intervalId');

    expect(window.scroll).not.toBeCalled();

    // fast forward one timer call
    jest.runOnlyPendingTimers();
    expect(setInterval).toHaveBeenCalledTimes(1);
    expect(window.scroll).toHaveBeenCalledTimes(1);
    expect(window.scroll).toHaveBeenCalledWith(0, 1000 - 50);
    expect(window.pageYOffset).toBe(1000 - 50);

    // fast forward until the end
    jest.runAllTimers();
    expect(window.scroll).toHaveBeenCalledTimes(20);
    expect(window.scroll).toHaveBeenLastCalledWith(0, 0);
    expect(window.pageYOffset).toBe(0);
    expect(clearInterval).toHaveBeenCalledTimes(1);
    expect(clearInterval).toHaveBeenCalledWith(intervalId);
    expect(wrapper.state('intervalId')).toBe(0);
  });
});
