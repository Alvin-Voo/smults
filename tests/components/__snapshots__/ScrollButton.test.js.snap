// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<Button
  as="button"
  className="scrollButton"
  icon="angle double up"
  onClick={[Function]}
  role="button"
  title="Back to top"
/>
`;
